src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -553,42 +557,31 @@ void recordWriteAfterQueryExecution() throws SQLException {
     }
 
     public Statement getReadStatement(RouteDecisionBuilder decisionBuilder) {
+        decisionBuilder.isSqlWrite(false);
         if (connectionProvider.getState().equals(MAIN)) {
             decisionBuilder.reason(MAIN_CONNECTION_REUSE);
             connectionProvider.getStateDecision().ifPresent(decisionBuilder::cause);
             return prepareWriteStatement(decisionBuilder);
         }
         final String sql = decisionBuilder.getSql();
-        isWriteOperation = sqlFunction.isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
+        isWriteOperation = sqlOperationType.isWriteOpearation(sql);
         if (isWriteOperation) {
-            decisionBuilder.reason(WRITE_OPERATION);
+            decisionBuilder.reason(WRITE_OPERATION).isSqlWrite(true);
             return prepareWriteStatement(decisionBuilder);
         }
 
-        if (isSelectForUpdate(sql)) {
-            decisionBuilder.reason(LOCK);
+        if (sqlOperationType.isSelectForUpdate(sql)) {
+            decisionBuilder.reason(LOCK).isSqlWrite(true);
             return prepareWriteStatement(decisionBuilder);
         }
-
         setCurrentStatement(getCurrentStatement() != null ? getCurrentStatement() : readStatement.get(decisionBuilder));
         performOperations();
         return getCurrentStatement();
     }
 
-    private boolean isSelectForUpdate(String sql) {
-        return sql != null && (sql.endsWith("for update") || sql.endsWith("FOR UPDATE"));
-    }
-
-    private boolean isUpdate(String sql) {
-        return sql != null && (sql.startsWith("update") || sql.startsWith("UPDATE"));
-    }
-
-    private boolean isDelete(String sql) {
-        return sql != null && (sql.startsWith("delete") || sql.startsWith("DELETE"));
-    }
-
     protected Statement getWriteStatement(RouteDecisionBuilder decisionBuilder) {
         isWriteOperation = true;
+        decisionBuilder.isSqlWrite(true);
```

Comment:
Here `String sql` isn't checked, which breaks the behavioral contract of `@return whether sqlQuery is recognized as writeOperation`

Reply:
Ha! Good catch. That's true! Cause `db-replica` recognizes whether query was write/read based on the method that call it. (based on `Statement` API) 
It works for almost all cases except from execute, which is mentioned here https://github.com/atlassian-labs/db-replica/issues/53 and It works for us as currently as our only product currently doesn't used execute() with read queries. Anyway this will be solved and isWrite will be set on compile time
https://github.com/atlassian-labs/db-replica/pull/49#discussion_r588348234